### PR TITLE
fix LAPACK.ormlq! and postmultiplication with LQPackedQ

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request fixes `LAPACK.ormlq!`'s shape checks, consequently fixing the bug in postmultiplication with `LQPackedQ`s identified in JuliaLang/LinearAlgebra.jl#469.  ~This pull request also makes `LQPackedQ`'s context-dependent postmultiplication behavior (see https://github.com/JuliaLang/julia/pull/23785#discussion_r140058503) consistent across all existing out-of-place multiplication methods (`*`, `A[c]_mul_B[c]`)~ (edit: botched) and tests those methods' behavior. Best!

(The first commit is JuliaLang/julia#23785, which I will rebase out.)

### CODE REVIEW

File: base/linalg/lq.jl [hunk 1]
```diff
@@ -159,39 +164,60 @@ for (f1, f2) in ((:A_mul_Bc, :A_mul_B!),
...
+# out-of-place right-application of LQPackedQs
+# unlike their in-place equivalents, these methods: (1) check whether the applied-to
+# matrix's (A's) appropriate dimension (columns for A_*, rows for Ac_*) matches the
+# number of columns (nQ) of the LQPackedQ (Q), as the underlying LAPACK routine (ormlq)
+# treats the implicit Q as its (nQ-by-nQ) square form; and (2) if the preceding dimensions
+# do not match, these methods check whether the appropriate dimension of A instead matches
+# the number of rows of the matrix of which Q is a factor (i.e. size(Q.factors, 1)),
+# and if so zero-extends A as necessary for check (1) to pass (if possible).
+*(A::StridedVecOrMat, Q::LQPackedQ) = _A_mul_Bq(A_mul_B!, A, Q)
+A_mul_Bc(A::StridedVecOrMat, Q::LQPackedQ) = _A_mul_Bq(A_mul_Bc!, A, Q)
```
Comment (Sacha0) [inline base/linalg/lq.jl]: This method does not require `LQPackedQ`'s context-dependent multiplication behavior, as `LQPackedQ`'s second dimension is not context-dependent.

File: base/linalg/lq.jl [hunk 2]
```diff
@@ -159,39 +164,60 @@ for (f1, f2) in ((:A_mul_Bc, :A_mul_B!),
...
-        function ($f1)(A::StridedMatrix, B::LQPackedQ)
-            TAB = promote_type(eltype(A), eltype(B))
-            AA = similar(A, TAB, (size(A, 2), size(A, 1)))
-            adjoint!(AA, A)
-            return ($f2)(AA, B)
-        end
+    return A_mul_Bop!(C, convert(AbstractMatrix{TR}, Q))
+end
+Ac_mul_B(A::StridedMatrix, Q::LQPackedQ) = _Ac_mul_Bq(A_mul_B!, A, Q)
+Ac_mul_Bc(A::StridedMatrix, Q::LQPackedQ) = _Ac_mul_Bq(A_mul_Bc!, A, Q)
```
Comment (Sacha0) [inline base/linalg/lq.jl]: This method also does not require `LQPackedQ`'s context-dependent multiplication behavior, as `LQPackedQ`'s second dimension is not context-dependent.

**Review by Sacha0 — COMMENTED** [review]
The `A_mul_Bc(C, Q)` and `Ac_mul_Bc(C, Q)` methods in this pull request do not require `LQPackedQ`'s context-dependent multiplication behavior as `LQPackedQ`'s second dimension is not context-dependent. Revisions inbound tomorrow. Thanks!

**Review by andreasnoack — APPROVED** [review]
Great work and thanks for adding all the comments.

**Comment by Sacha0** [comment]
Thanks Andreas! I will prepare a followup pull request with the touchups mentioned above :).

**Pull request merged**